### PR TITLE
Compute Delaunay triangulation using PDAL

### PR DESCRIPTION
**Pull request by jpolchlo**
This PR shifts away from the GT Delaunay triangulator in favor of the one provided by PDAL.  We also borrow the facilities from GT to rasterize this new triangulation.

### CODE REVIEW

File: pointcloud/src/main/scala/geotrellis/pointcloud/raster/ept/DEMRasterSource.scala [hunk 1]
```diff
@@ -84,16 +84,16 @@ case class DEMRasterSource(
...
     val expression = ReadEpt(
       filename   = eptSource,
       resolution = gridExtent.cellSize.resolution.some,
       bounds     = s"([${bnds.xmin}, ${bnds.ymin}], [${bnds.xmax}, ${bnds.ymax}])".some,
       threads    = threads
-    )
+    ) ~ FilterDelaunay()
 
-    logger.debug(expression.toPipelineConstructor.asJson.spaces4)
+    //logger.debug(expression.toPipelineConstructor.asJson.spaces4)
```
Comment (pomadchin) [inline pointcloud/src/main/scala/geotrellis/pointcloud/raster/ept/DEMRasterSource.scala]: ah, it is a pipeline constructor already; so 

```scala
logger.debug(expression.asJson.spaces4)
```

would work; I use it to print out the pipeline generated by the geotrellis server.

File: pointcloud/src/main/scala/geotrellis/pointcloud/raster/ept/DEMRasterSource.scala [hunk 2]
```diff
@@ -102,16 +102,124 @@ case class DEMRasterSource(
         pipeline.execute
 
         val pointViews = pipeline.getPointViews().asScala.toList
-        val viewSizes = pointViews.map(_.length)
-        val (nCoords, starts) = viewSizes.foldLeft(0 -> List.empty[Int]) { case ((acc, cum), v) => (v + acc, cum :+ acc) }
-        val coords = Array.ofDim[Coordinate](nCoords)
-        viewSizes.zip(starts).zip(pointViews).foreach { case ((n, z), pv) =>
-          val pc = pv.getPointCloud(DimType.X, DimType.Y, DimType.Z)
-          cfor(0)(_ < n, _ + 1) { i => coords(z + i) = pc.getCoordinate(i) }
+        assert(pointViews.length == 1, "Triangulation pipeline should have single resulting point view")
```
Comment (pomadchin) [inline pointcloud/src/main/scala/geotrellis/pointcloud/raster/ept/DEMRasterSource.scala:105]: Oops, I think we can not do that in the semantics of the current read function. 

The type signature of this read function implies that `None` is an expected behavior and we don't expect it to throw. Mb just adding a `logger.warn` instead of throwing here would be enough, otherwise `Option[Raster]` signature is completely broken.

For the context: in the future RasterSources would be abstracted over the effect type: [GeoTiffRasterSource example](https://github.com/geotrellis/geotrellis-contrib/blob/develop/vlm/src/main/scala/geotrellis/contrib/vlm/effect/geotiff/GeoTiffRasterSource.scala#L74) and a [RasterSource interface example](https://github.com/geotrellis/geotrellis-contrib/blob/develop/vlm/src/main/scala/geotrellis/contrib/vlm/effect/RasterSourceF.scala), so users would have control over the effect (Either / Option / IO / Task / etc).

File: pointcloud/src/main/scala/geotrellis/pointcloud/raster/ept/DEMRasterSource.scala [hunk 3]
```diff
@@ -102,16 +102,124 @@ case class DEMRasterSource(
...
+                }
+
+                tile.setDouble(col, row, z)
+              }
+
+              scanx += w
+            }
+
+            scany += h
+          }
```
Comment (pomadchin) [inline pointcloud/src/main/scala/geotrellis/pointcloud/raster/ept/DEMRasterSource.scala]: How do you feel about moving this code into some `TriangleRasterizer`?

File: pointcloud/src/main/scala/geotrellis/pointcloud/raster/ept/DEMRasterSource.scala [hunk 4]
```diff
@@ -102,16 +102,124 @@ case class DEMRasterSource(
...
+        val tris = pointViews(0).getTriangularMesh()
+
+        val w = gridExtent.cellwidth
+        val h = gridExtent.cellheight
+        val cols = bounds.width
+        val rows = bounds.height
+        val tile = DoubleArrayTile.empty(cols.toInt, rows.toInt)
+
+        def rasterizeTriangle(tri: Triangle): Unit = {
+          val (a,b,c) = (tri.a, tri.b, tri.c)
```
Comment (pomadchin) [inline pointcloud/src/main/scala/geotrellis/pointcloud/raster/ept/DEMRasterSource.scala]: Btw, [Triangle](https://github.com/PDAL/java/blob/2.1.4/core/src/main/scala/io/pdal/Triangle.scala#L37-L46) is isomorphic to tuple3.

**Review by pomadchin — CHANGES_REQUESTED** [review]
Looks almost perfect, but mb some code cleanup would be good to do before merging it.

File: pointcloud/src/main/scala/geotrellis/pointcloud/raster/ept/DEMRasterSource.scala (hunk 3 above)
Comment (jpolchlo) [reply pointcloud/src/main/scala/geotrellis/pointcloud/raster/ept/DEMRasterSource.scala]: That is possible, but we'd want to check performance of that?  This rasterizer pulls some amount of state from the surrounding scope.  This was done originally because of the use of the half-edge mesh.  I'd be concerned about increasing the call times at all of a function that will be run hundreds of thousands of times.  We declared this function inline in the DelaunayRasterizer, which was tuned for performance, so I followed that model.  We can come back and test whether it makes a difference later, but for now, I'd say leave it in place.

File: pointcloud/src/main/scala/geotrellis/pointcloud/raster/ept/DEMRasterSource.scala (hunk 2 above)
Comment (jpolchlo) [reply pointcloud/src/main/scala/geotrellis/pointcloud/raster/ept/DEMRasterSource.scala:105]: In this case, I'd rather not lose the explicit check.  This is a weird edge case that we don't know much about.  Rather than a whisper, I want this code to scream when we find a test case of this.

File: pointcloud/src/main/scala/geotrellis/pointcloud/raster/ept/DEMRasterSource.scala (hunk 4 above)
Comment (jpolchlo) [reply pointcloud/src/main/scala/geotrellis/pointcloud/raster/ept/DEMRasterSource.scala]: And yet,
```scala
val (a,b,c) = tri
```
Failed to compile.  Probably better to use
```scala
val Triangle(a,b,c) = tri
```

File: pointcloud/src/main/scala/geotrellis/pointcloud/raster/ept/DEMRasterSource.scala (hunk 3 above)
Comment (pomadchin) [reply pointcloud/src/main/scala/geotrellis/pointcloud/raster/ept/DEMRasterSource.scala]: I don't follow your implications, but I'll merge it as it is.

**Review by pomadchin — APPROVED** [review]
Mering as it is, after closing https://github.com/geotrellis/geotrellis-pointcloud/issues/37 there would be a follow up refactor to move all the triangulation logic away from the RasterSource.
